packages/rocketchat-autotranslate/server/autotranslate.js

```diff
@@ -220,38 +206,92 @@ class AutoTranslate {
 		return message;
 	}
 
-	getSupportedLanguages(target) {
-		if (this.enabled && this.apiKey) {
-			if (this.supportedLanguages[target]) {
-				return this.supportedLanguages[target];
-			}
+	/**
+	 * Registers afterSaveMessage call back for active service provider
+	 * @protected
+	 * @param {string} provider
+	 */
+	_registerAfterSaveMsgCallBack(provider) {
+		RocketChat.callbacks.add('afterSaveMessage', this.translateMessage.bind(this), RocketChat.callbacks.priority.MEDIUM, provider);
+	}
 
-			let result;
-			const params = { key: this.apiKey };
-			if (target) {
-				params.target = target;
-			}
+	/**
+	 * De-register afterSaveMessage call back for active service provider
+	 * @protected
+	 * @param {string} provider
+	 */
+	_unRegisterAfterSaveMsgCallBack(provider) {
+		RocketChat.callbacks.remove('afterSaveMessage', provider);
+	}
 
-			try {
-				result = HTTP.get('https://translation.googleapis.com/language/translate/v2/languages', { params });
-			} catch (e) {
-				if (e.response && e.response.statusCode === 400 && e.response.data && e.response.data.error && e.response.data.error.status === 'INVALID_ARGUMENT') {
-					params.target = 'en';
-					target = 'en';
-					if (!this.supportedLanguages[target]) {
-						result = HTTP.get('https://translation.googleapis.com/language/translate/v2/languages', { params });
-					}
-				}
-			} finally {
-				if (this.supportedLanguages[target]) {
-					return this.supportedLanguages[target];
-				} else {
-					this.supportedLanguages[target || 'en'] = result && result.data && result.data.data && result.data.data.languages;
-					return this.supportedLanguages[target || 'en'];
-				}
-			}
+	/**
+	 * Returns metadata information about the service provider
+	 * @abstract
+	 * @private
+	 * @returns {object}
+	 */
+	_getProviderMetadata() {
+		SystemLogger.warn('must be implemented by subclass!', '_getProviderMetadata');
+	}
+
+
+	/**
+	 * Returns necessary settings information about the translation service provider.
+	 * @abstract
+	 * @private
+	 * @param {string} target
+	 * @returns {object}
+	 */
+	_getSupportedLanguages(target) {
+		SystemLogger.warn('must be implemented by subclass!', '_getSupportedLanguages', target);
+	}
+
+	/**
+	 * Send Request REST API call to the service provider.
+	 * @abstract
+	 * @public
+	 * @param {object} message
+	 * @param {object} targetLanguages
+	 * @return {object}
+	 */
+	_sendRequestTranslateMessage(message, targetLanguages) {
+		SystemLogger.warn('must be implemented by subclass!', '_sendRequestTranslateMessage', message, targetLanguages);
+	}
+
+	/**
+	 * Send Request REST API call to the service provider.
```

Comment:
Particularly for the abstract methods, the description is important. Shall this translate the attachments or only the attachment's description?